multibody/multibody_solvers.cc

```diff
@@ -0,0 +1,171 @@
+#include "multibody/multibody_solvers.h"
+#include "multibody/multibody_utils.h"
+
+namespace dairlib {
+namespace multibody {
+
+using Eigen::VectorXd;
+using drake::multibody::MultibodyPlant;
+using drake::solvers::Binding;
+using drake::solvers::Constraint;
+using drake::solvers::VectorXDecisionVariable;
+using drake::systems::Context;
+using drake::MatrixX;
+using drake::VectorX;
+using solvers::NonlinearConstraint;
+
+template <typename T>
+MultibodyProgram<T>::MultibodyProgram(const MultibodyPlant<T>& plant,
+    const KinematicEvaluatorSet<T>& evaluators)
+    : drake::solvers::MathematicalProgram(),
+    plant_(plant),
+    evaluators_(evaluators),
+    context_(std::shared_ptr<Context<T>>(
+        plant_.CreateDefaultContext().release())) {}
+
+template <typename T>
+VectorXDecisionVariable MultibodyProgram<T>::AddPositionVariables() {
+  DRAKE_DEMAND(q_.size() == 0);
+  q_ = NewContinuousVariables(plant_.num_positions(), "q");
+  return q_;
+}
+
+template <typename T>
+VectorXDecisionVariable MultibodyProgram<T>::AddInputVariables() {
+  DRAKE_DEMAND(u_.size() == 0);
+  u_ = NewContinuousVariables(plant_.num_actuators(), "u");
+  return u_;
+}
+
+template <typename T>
+VectorXDecisionVariable MultibodyProgram<T>::AddConstraintForceVariables() {
+  DRAKE_DEMAND(lambda_.size() == 0);
+  lambda_ = NewContinuousVariables(evaluators_.count_full(), "lambda");
+  return lambda_;
+}
+
+template <typename T>
+Binding<Constraint> MultibodyProgram<T>::AddKinematicConstraint(
+    const VectorXDecisionVariable& q) {
+  auto constraint = std::make_shared<KinematicPositionConstraint<T>>(
+        plant_, evaluators_, context_);
+  return AddConstraint(constraint, q);
+}
+
+template <typename T>
+Binding<Constraint> MultibodyProgram<T>::AddFixedPointConstraint(
+    const VectorXDecisionVariable& q, const VectorXDecisionVariable& u,
+    const VectorXDecisionVariable& lambda) {
+  DRAKE_DEMAND(lambda.size() == evaluators_.count_full());
+  auto constraint = std::make_shared<FixedPointConstraint<T>>(
+        plant_, evaluators_, context_);
+  return AddConstraint(constraint, {q, u, lambda});
+}
+
+template <typename T>
+void MultibodyProgram<T>::AddJointLimitConstraints(VectorXDecisionVariable q) {
+  for (int i = 0; i < plant_.num_joints(); i++) {
+    const auto& joint = plant_.get_joint(drake::multibody::JointIndex(i));
+    if (joint.num_positions() > 0) {
+      auto q_joint = q.segment(joint.position_start(), joint.num_positions());
+      AddConstraint(q_joint <= joint.position_upper_limits());
```

Comment:
Just to confirm, do the joint limits work in the case of continuous time MBP? I remember seeing warnings about joint limits are not supportet in this case.

Reply:
You're right, they are not used when _simulating_ the continuous time MBP. However, they still exist and can be referenced.

Reply:
Got it!